### PR TITLE
[Android] Fix mEditable NullPointerException in TextInputPlugin

### PR DESCRIPTION
**Pull request by eggfly**
Fix:
* https://github.com/flutter/flutter/issues/94794
* https://github.com/flutter/flutter/issues/92241

Our app also faced this problem in our internal crash report system, see the root cause from below comments.

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 1]
```diff
@@ -559,7 +559,9 @@ void clearTextInputClient() {
       // gained focus as part of the same focus event.
       return;
     }
-    mEditable.removeEditingStateListener(this);
+    if (mEditable != null) {
+      mEditable.removeEditingStateListener(this);
```
Comment (blasten) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:563]: this could have a test

**Comment by eggfly** [comment]
I thought a lot about it, there are two MethodChannel invokes to Platform thread called: `'TextInput.setClient'` and `'TextInput.setEditingState`. the `setEditingState` is always called after `_attach`.


https://github.com/flutter/flutter/blob/ffcd32ebb68fbfe3c9788a3bf1a09d3ebe28887f/packages/flutter/lib/src/services/text_input.dart#L1495

```dart
  /// This method actually notifies the embedding of the client. It is utilized
  /// by [attach] and by [_handleTextInputInvocation] for the
  /// `TextInputClient.requestExistingInputState` method.
  void _attach(TextInputConnection connection, TextInputConfiguration configuration) {
    assert(connection != null);
    assert(connection._client != null);
    assert(configuration != null);
    assert(_debugEnsureInputActionWorksOnPlatform(configuration.inputAction));
    _channel.invokeMethod<void>(
      'TextInput.setClient',
      <dynamic>[ connection._id, configuration.toJson() ],
    );
    _currentConnection = connection;
    _currentConfiguration = configuration;
  }
```

https://github.com/flutter/flutter/blob/ffcd32ebb68fbfe3c9788a3bf1a09d3ebe28887f/packages/flutter/lib/src/services/text_input.dart#L1666
```dart
  void _setEditingState(TextEditingValue value) {
    assert(value != null);
    _channel.invokeMethod<void>(
      'TextInput.setEditingState',
      value.toJSON(),
    );
  }
```

https://github.com/flutter/engine/blob/2962099077b36704071802ea5595f4a016a1c214/shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java#L408

```java
  @VisibleForTesting
  void setTextInputClient(int client, TextInputChannel.Configuration configuration) {
    // Call notifyViewExited on the previous field.
    notifyViewExited();
    inputTarget = new InputTarget(InputTarget.Type.FRAMEWORK_CLIENT, client);

    if (mEditable != null) {
      mEditable.removeEditingStateListener(this);
    }
    mEditable =
        new ListenableEditingState(
            configuration.autofill != null ? configuration.autofill.editState : null, mView);   // <---- here 
    // more code ..
  }
```


So there's impossible to leave the `mEditable` member in `TextInputPlugin` class.
UNLESS: there's another `FlutterView` object?

https://github.com/flutter/engine/blob/e11a2579a0e941112de02be988724529000b44c1/shell/platform/android/io/flutter/embedding/android/FlutterView.java#L1080
```
  public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
    Log.v(TAG, "Attaching to a FlutterEngine: " + flutterEngine);
    if (isAttachedToFlutterEngine()) {
      if (flutterEngine == this.flutterEngine) {
        // We are already attached to this FlutterEngine
        Log.v(TAG, "Already attached to this engine. Doing nothing.");
        return;
      }

      // Detach from a previous FlutterEngine so we can attach to this new one.
      Log.v(
          TAG,
          "Currently attached to a different engine. Detaching and then attaching"
              + " to new engine.");
      detachFromFlutterEngine();
    }

    this.flutterEngine = flutterEngine;

    // Instruct our FlutterRenderer that we are now its designated RenderSurface.
    FlutterRenderer flutterRenderer = this.flutterEngine.getRenderer();
    isFlutterUiDisplayed = flutterRenderer.isDisplayingFlutterUi();
    renderSurface.attachToRenderer(flutterRenderer);
    flutterRenderer.addIsDisplayingFlutterUiListener(flutterUiDisplayListener);

    // Initialize various components that know how to process Android View I/O
    // in a way that Flutter understands.
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      mouseCursorPlugin = new MouseCursorPlugin(this, this.flutterEngine.getMouseCursorChannel());
    }
    textInputPlugin =
        new TextInputPlugin(
            this,
            this.flutterEngine.getTextInputChannel(),
            this.flutterEngine.getPlatformViewsController());
```


```java

  @VisibleForTesting
  void setTextInputClient(int client, TextInputChannel.Configuration configuration) {
    // Call notifyViewExited on the previous field.
    notifyViewExited();
    inputTarget = new InputTarget(InputTarget.Type.FRAMEWORK_CLIENT, client);

    if (mEditable != null) {
      mEditable.removeEditingStateListener(this);
    }
    mEditable =
        new ListenableEditingState(
            configuration.autofill != null ? configuration.autofill.editState : null, mView);
    this.configuration = configuration;
    updateAutofillConfigurationIfNeeded(configuration);

    // setTextInputClient will be followed by a call to setTextInputEditingState.
    // Do a restartInput at that time.
    mRestartInputPending = true;
    unlockPlatformViewInputConnection();
    lastClientRect = null;
    mEditable.addEditingStateListener(this);
  }

```


If we try to attach an old engine (reuse engine) to a new `FlutterView` instance and call setTextInputEditingState or call clearTextInputClient, the `mEditable` maybe null. It will crash.

I think we can create a `mEditable` object in the constructor of `TextInputPlugin`? Give it a default client_id and `mView = FlutterView.this` ? @LongCatIsLooong @blasten

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 1 above)
Comment (eggfly) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Done~

File: shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java [hunk 2]
```diff
@@ -55,10 +55,10 @@ void didChangeEditingState(
 
   // The View is only used for creating a dummy BaseInputConnection for setComposingRegion. The View
   // needs to have a non-null Context.
-  public ListenableEditingState(TextInputChannel.TextEditState initalState, View view) {
+  public ListenableEditingState(TextInputChannel.TextEditState initialState, View view) {
```
Comment (blasten) [inline shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java:58]: nit: add nullable/non-null annotations

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 3]
```diff
@@ -66,6 +66,8 @@ public TextInputPlugin(
       @NonNull TextInputChannel textInputChannel,
       @NonNull PlatformViewsController platformViewsController) {
     mView = view;
+    // Create a default object.
+    mEditable = new ListenableEditingState(null, mView);
```
Comment (blasten) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:70]: nit: replace `mEditable` annotation `non-null/nullable`.

**Review by blasten — APPROVED** [review]
LGTM % nits

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 3 above)
Comment (eggfly) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:70]: done~

File: shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java (hunk 2 above)
Comment (eggfly) [reply shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java]: done~

**Pull request merged**